tango/common/logging.py

```diff
@@ -1,72 +1,196 @@
+main process (the tango CLI does this for you), and you've passed the logging queue
+as an argument to your worker functions. You can get this queue by calling :func:`get_logging_queue()`.
+
+For example,
+
+.. code-block::
+
+    import logging
+    import multiprocessing as mp
+
+    from tango import Step
+    from tango.common.logging import initialize_worker_logging, get_logging_queue
+
+    @Step.register("multiprocessing_step")
+    class MultiprocessingStep(Step):
+        def run(self, num_proc: int = 2) -> bool:  # type: ignore
+            workers = []
+            for i in range(num_proc):
+                worker = mp.Process(target=_worker_function, args=(i, get_logging_queue()))
+                workers.append(worker)
+                worker.start()
+            for worker in workers:
+                worker.join()
+            return True
+
+
+    def _worker_function(worker_id: int, logging_queue: mp.Queue):
+        initialize_worker_logging(worker_id, logging_queue)
+        logger = logging.getLogger(MultiprocessingStep.__name__)
+        logger.info("Hello from worker %d!", worker_id)
+
+"""
+
 import logging
+import logging.handlers
+import multiprocessing as mp
 import os
+import sys
+import threading
+from contextlib import contextmanager
 from typing import Optional
 
 import click
 
+from .aliases import PathOrStr
+from .exceptions import SigTermReceived
 from .util import _parse_bool
 
+FILE_FRIENDLY_LOGGING: bool = _parse_bool(os.environ.get("FILE_FRIENDLY_LOGGING", False))
+"""
+If this flag is set to ``True``, we remove special styling characters from log messages,
+add newlines to tqdm output even on an interactive terminal, and we slow
+down tqdm's output to only once every 10 seconds.
+
+By default, it is set to ``False``. It can be changed by setting the corresponding environment
+variable (``FILE_FRIENDLY_LOGGING``) or field in a :class:`~tango.__main__.TangoGlobalSettings`
+file (``file_friendly_logging``) to "true" or "false",
+or from the command line with the ``--file-friendly-logging`` flag.
+For example,
+
+.. code-block::
+
+    $ tango --file-friendly-logging run ...
+
+"""
+
+TANGO_LOG_LEVEL: Optional[str] = os.environ.get("TANGO_LOG_LEVEL", None)
+"""
+The log level to use globally. The value can be set from the corresponding environment variable
+(``TANGO_LOG_LEVEL``) or field in a :class:`~tango.__main__.TangoGlobalSettings` file (``log_level``),
+or from the command line with the ``--log-level`` option.
+Possible values are "debug", "info", "warning", or "error" (not case sensitive).
+For example,
+
+.. code-block::
+
+    $ tango --log-level info run ...
+
+"""
+
+_LOGGING_QUEUE: Optional[mp.Queue] = None
+"""
+Used to send log records from worker processes back to the main logging thread.
+"""
+
+_LOGGING_THREAD: Optional[threading.Thread] = None
+"""
+Thread used for logging records from worker processes.
+"""
```

Comment:
This queue and thread business was inspired by https://docs.python.org/dev/howto/logging-cookbook.html#logging-to-a-single-file-from-multiple-processes

Reply:
It's necessary to safely write logs from worker processes to the same file.